### PR TITLE
Expose Oauth2 Endpoint Parameters to cli

### PR DESCRIPTION
**Pull request by mhill-holoplot**
The user may which to control the endpoint parameters for instance to set the audience when requesting an access token. Exposing the parameters as a map allows for additional use cases without requiring modification.

### CODE REVIEW

File: lib/flagutil/map.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+func parseMapValue(s string) (string, string) {
+	kv := strings.SplitN(s, ":", 2)
+	return kv[0], kv[1]
+}
+
+// NewMapString returns a new MapString with the given name and description.
+func NewMapString(name, description string) *MapString {
+	description += "\nSupports multiple flags with the following syntax: -name=key:value"
```
Comment (tenmozes) [inline lib/flagutil/map.go:33]: as far you know the name why don't use it in description
Supports multiple flags with the following syntax: -name=key:value -> Supports multiple flags with the following syntax: -%s=key:value (where %s -> name variable)

File: lib/flagutil/map.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import (
+	"flag"
+	"fmt"
+	"strings"
+)
+
+type MapString map[string]string
+
+func (m *MapString) String() string {
+	return fmt.Sprintf("%v", *m)
```
Comment (tenmozes) [inline lib/flagutil/map.go:12]: if nill return "" ?

File: lib/flagutil/map.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	for _, pair := range parseArrayValues(value) {
+		key, value := parseMapValue(pair)
+		(*m)[key] = value
+	}
+	return nil
+}
+
+func parseMapValue(s string) (string, string) {
+	kv := strings.SplitN(s, ":", 2)
+	return kv[0], kv[1]
```
Comment (tenmozes) [inline lib/flagutil/map.go:28]: this panic if no `:` see https://go.dev/play/p/9aA8Dk9qHg1
which is not bad but we can panic more explicitly

File: lib/flagutil/map.go (hunk 2 above)
Comment (mhill-holoplot) [reply lib/flagutil/map.go:12]: yeah I guess I should handle null there was being a bit lazy.

File: lib/flagutil/map.go (hunk 3 above)
Comment (mhill-holoplot) [reply lib/flagutil/map.go:28]: True ideally one would present an error to the user that makes sense. Is there are preferred way to do that here?

Comment (mhill-holoplot) [reply lib/flagutil/map.go:28]: Actually just spotted I can through an error all the way back up the stack which is nicer.

**Comment by mhill-holoplot** [comment]
Fixed the lint issue now.

**Review by valyala — APPROVED** [review]

**Comment by valyala** [comment]
@mhill-holoplot , thanks for adding the ability to provide Oauth2 Endpoint parameters via `-remoteWrite.oauth2.endpointParams` command-line flag! This feature will be included in the next release!

**Comment by valyala** [comment]
@mhill-holoplot , please see the follow-up commit 2f41cdbb9af0a9ba3baa33a9b1055866d239895a , which adds `*.oauth2.endpointParams` command-line flags to all the places in [vmalert](https://docs.victoriametrics.com/vmalert.html), which support OAuth2 configuration.

The follow-up commit also allows configuring independent OAuth2 endpoint params per each `-remoteWrite.url`, since the previous implementation was sharing a single set of endpoint params across all the `-remoteWrite.url` flags. For example, the following config sets different OAuth2 endpoint params per each `-remoteWrite.url`:

```
./vmagent \
  -remoteWrite.url=http://remote-storage-1/api/v1/write \
  -remoteWrite.oauth2.endpointParams='{"foo":"bar","a":"b"}' \
  -remoteWrite.oauth2.clientSecret=secret \
  -remoteWrite.oauth2.clientID=abc \
  -remoteWrite.oauth2.tokenUrl=http://oauth2-host-1/token/url \
  \
  -remoteWrite.url=http://remote-storage-2/api/v1/write \
  -remoteWrite.oauth2.endpointParams='{"x":"y"}' \
  -remoteWrite.oauth2.clientSecret=secret \
  -remoteWrite.oauth2.clientID=def \
  -remoteWrite.oauth2.tokenUrl=http://oauth2-host-2/token/url
```

**Comment by valyala** [comment]
FYI, this pull request has been included in [v1.97.0](https://docs.victoriametrics.com/CHANGELOG.html).

**Pull request merged**